### PR TITLE
Add support for the `tokenizers` package

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This change adds support for the `tokenizers` package. We started implementing support for reading `tokenizer.json` files. However, fully supporting this format is currently not our focus, so instead we use the `tokenizers` package to support these tokenizers.

To support this, the class hierarchy needs to be restructed a bit:

* The original `Tokenizer` class is split into two classes: `TokenizerBase` defines the interface for tokenizers. `LegacyTokenizer` implements the interface for our original tokenizers.
* The new `Tokenizer` class it the `tokenizers`-based tokenizer. This tokenizer has a prominent name, because going forward it will be the preferred/default tokenizer.

### Types of change

Feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/tokenization/tokenizer.py [hunk 1]
```diff
@@ -73,6 +83,172 @@ def padded_tensor(self, *, padding_id: int, pad_left: bool = False):
...
+            seq_pieces = []
+
+            for chunk in seq:
+                if isinstance(chunk, MergedSpecialPieceChunk):
+                    piece_id = self.tokenizer.token_to_id(chunk.piece)
+                    seq_ids.append(piece_id)
+                    seq_pieces.append(chunk.piece)
+                else:
+                    encoding = self.tokenizer.encode(
+                        chunk.text,  # add_special_tokens=False
```
Comment (shadeMe) [inline curated_transformers/tokenization/tokenizer.py:204]: Intentionally commented out?

File: setup.cfg [hunk 2]
```diff
@@ -16,6 +16,7 @@ python_requires = >=3.8
 install_requires =
     curated-tokenizers>=0.0.7,<0.1.0
     huggingface-hub>=0.14
+    tokenizers
```
Comment (shadeMe) [inline setup.cfg:19]: No pins on this requirement?

File: curated_transformers/tokenization/tokenizer.py [hunk 3]
```diff
@@ -73,6 +83,172 @@ def padded_tensor(self, *, padding_id: int, pad_left: bool = False):
...
+
+        decoded = [clean_up_decoded_string_like_hf(string) for string in decoded]
+
+        return decoded
+
+    def encode(
+        self, input: Union[Iterable[InputChunks], Iterable[str]]
+    ) -> PiecesWithIds:
+        input = cast(Union[List[InputChunks], List[str]], list(input))
+        if isinstance(input[0], str):
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/tokenization/tokenizer.py:174]: I find these decision difficult, one could argue that this is the role of type checking. But in this case the errors are really ugly if you use incorrect input, so I added validation.

File: curated_transformers/tokenization/tokenizer.py (hunk 1 above)
Comment (danieldk) [reply curated_transformers/tokenization/tokenizer.py]: I think that was probably during debugging, removed comment.

File: setup.cfg (hunk 2 above)
Comment (danieldk) [reply setup.cfg]: Set the lower version to the current version.

**Review by shadeMe — APPROVED** [review]

**Pull request merged**